File: pattern/observer.cc
//written by lissa
//date 2015 07 15
//
//pattern Subject and Observer.


//compile:g++ <this file name>


#include<iostream>
using namespace std;
#include<vector>
class Observer;
class ConcreteSubject;
class ConcreteObserver;

class Observer{
public:
	virtual void update(string,string)=0;
};

class Subject{
public:
	void attach(Observer* observer){
		m_vector.push_back(observer);
	};
	void detach(Observer*observer){
		vector<Observer*>::iterator it=m_vector.begin();
		for(;it!=m_vector.end();it++){
			if(*it==observer)
				m_vector.erase(it);
		}
	};
	void notify(string key,string value){
		vector<Observer*>::iterator it=m_vector.begin();
		for(;it!=m_vector.end();it++){
			(*it)->update(key,value);
		}
	};
protected:
	vector<Observer*> m_vector;
private:
};

class ConcreteSubject:public Subject{
public:
	void setName(string name){
		this->name=name;
		Subject::notify("name",this->name);
	};
private:
	string name;
};


class ConcreteObserver:public Observer{
public:
	virtual void update(string key,string value){
		cout<<"主题数据改变了"<<key<<":"<<value<<endl;
	};
};

int main(){
	Observer*ob1=new ConcreteObserver();
	Observer*ob2=new ConcreteObserver();
	Observer*ob3=new ConcreteObserver();

	ConcreteSubject *sub=new ConcreteSubject();

	sub->attach(ob1);
	sub->attach(ob2);
	sub->attach(ob3);
	
	sub->setName("lizhi");
	return 0;
}

File: pattern/observer.js
function inherit (superclass, subclass) {
    subclass.prototype = Object.create(superclass.prototype);
    subclass.prototype.constructor = subclass;
}

function Subject () {
    this.observers = new Array();

    Subject.prototype.attach = function (observer) {
        var index = this.observers.length;
        this.observers[index] = observer;
    };

    Subject.prototype.detach = function (observer) {
        for (var i = 0; i < this.observers.length; i++) {
            if (this.observers[i] === observer) {
                delete (this.observers[i]);
            }
        }
    };

    Subject.prototype.notify = function (key, value) {
        for (var i = 0; i < this.observers.length; i++) {
            if (this.observers[i] &&
                typeof this.observers[i].update === 'function') {
                this.observers[i].update(key, value);
            }
        }
    };
};

inherit(Subject, ConcreteSubject);
function ConcreteSubject () {
    Subject.call(this);
}


inherit(Subject, ConcreteSubjectX);
function ConcreteSubjectX () {
    Subject.call(this);
}



function Observer () {
    if (this.constructor === Observer) {
        throw new Error("Cannot instanciate abstract class");
    }

    Observer.prototype.update = function () {
        throw new Error("Cannot call abstract method");
    };
};


inherit(Observer, ConcreteObserver);
function ConcreteObserver () {
    Observer.call(this, arguments);

    ConcreteObserver.prototype.update = function (key, value) {
        console.log("hi--" + key + ":" + value);
    }
}


inherit(Observer, ConcreteObserverX);
function ConcreteObserverX () {
    Observer.call(this, arguments);

    ConcreteObserverX.prototype.update = function (key, value) {
        console.log("XXX--" + key + ":" + value);
    }
}

var cs = new ConcreteSubject();
var csx = new ConcreteSubjectX();
var co = new ConcreteObserver();
var co1 = new ConcreteObserver();
var co2 = new ConcreteObserverX();
csx.attach(co);
cs.attach(co);
cs.attach(co1);
cs.attach(co2);
/*
cs.notify("hello", "world");
cs.detach(co);
cs.notify("hellolll", "world");
*/

csx.notify("csx", "but");
csx.detach(co);
csx.notify("csx", "hsdkf");

